guides/hack/56-modules/04-inheritance.md

```diff
@@ -0,0 +1,67 @@
+# Module level visibility and inheritance
+
+## Inheriting methods and properties
+This section discusses the relationships between Hack's object hierarchy and `internal`. In general, override rules for internal class members are consistent with that of private and protected: you can only override a class member with a visibility that's at least as visible as the previous. 
+
+```hack
+module foo;
+class Foo {
+    internal function foo(): void {}
+
+    public function bar(): void {}
+
+    internal function baz(): void {}
+}
+class Bar extends Foo {
+    <<__Override>>
+    public function foo(): void {} // ok 
+
+    <<__Override>>
+    internal function bar(): void {} // error, overriding a public method with a lower visibility
+
+    <<__Override>>
+    private function baz(): void {} // ok
+}
+```
+
+Note that `internal` and `protected` do not encompass each other in terms of visibility. Therefore, you can never override an internal member with a protected one, or vice versa.
+
+The same rule applies when implementing an interface:
+```hack
+interface IFoo {
+    public function bar(): void;
+    internal function baz(): void;
+}
+class Foo implements IFoo {
+    // You must implement a public function in an interface with another public one
+    public function bar(): void {}
+    // You **can** implement an internal function in an interface with a public one
+    public function baz(): void {}
+}
+```
+
+
+## Inheriting toplevel symbols
+Unlike with methods and properties, within the same module, an internal class can override a public one, and vice versa. 
+
+```hack
+public class Bar {}
+internal class Foo extends Bar {}
+public class Baz extends Foo {}
+```
+You can think of this behavior as freely choosing which symbols in your module to export to your public API. Since the visibility of a toplevel entity only affects how it is statically referenced, no inheritance rules need to be applied when overriding them. Note that even if the toplevel symbol visibilities do not need to match any relationship, the individual members of those symbols do need to meet inheritance rules as discussed previously. 
```

Comment:
```suggestion
You can think of this behavior as freely choosing which symbols in your module to export to your public API. Since the visibility of a toplevel entity only affects how it is statically referenced, no inheritance rules need to be applied when overriding them.
```

This last sentence kinda sounds like something that's relevant when implementing modules, but less useful for users.